### ISSUE TITLE
vignette essay

### ISSUE THREAD

**Issue by johannes-titz** [shevandrin/qti#90]
Give a warning (or even error) when feedback+ or feedback- is provided because this does not make sense.

default could be made dependent on the number of words in feedback, which could also be used to define word limit and expected length. Not sure if a minimum makes sense.

max_strings -> strings_max
min_strings -> strings_min

Or is this inconvenient because qti implements it differently? if so, we can leave it as is. But I find strings_max and strings_min more intuitive, especially if searching for it.

data_allow_paste: is this a qti feature or opal?

__________

lets finalized the list of changes:

- [x] feedback+ or feedback- for Essay give an Error
- [x] max_strings -> words_max  and strings_min ->words_min
- [x] data_allow_paste  with warning that it works only on Opal
- [x] default value for words_max is number of words in model answer from feedback * 2
- [x] expected length is feedback * 1.5

- [x] cols="40" 

- [x] rows="5"

**Comment by shevandrin** [comment]
> Give a warning (or even error) when feedback+ or feedback- is provided because this does not make sense.

I vote for error, because it should be given clear content.

> default could be made dependent on the number of words in feedback, which could also be used to define word limit and expected length. Not sure if a minimum makes sense.

i dont get it, sorry)

> max_strings -> strings_max
> min_strings -> strings_min

yes, according to QTI `maxStrings [0..1]: [integer]`, that war transformed as ` max_string` as a slot name. If we have focus on final user, we can change the names

> data_allow_paste: is this a qti feature or opal?

yes, it is only for opal. I should have created inherited class  EssayOpal, but only due to one slot - i didnt) should i do it?

**Comment by johannes-titz** [comment]
> > Give a warning (or even error) when feedback+ or feedback- is provided because this does not make sense.
> 
> I vote for error, because it should be given clear content.

Agreed

> 
> > default could be made dependent on the number of words in feedback, which could also be used to define word limit and expected length. Not sure if a minimum makes sense.
> 
> i dont get it, sorry)

As an instructor I write the perfect answer in the feedback section. From this answer we can calculate expected length, maybe feedback-length * 1.5, and max as feedback-length*2. This is useful so that students know how much they need to write and to prevent too long answers.

> 
> > max_strings -> strings_max
> > min_strings -> strings_min
> 
> yes, according to QTI `maxStrings [0..1]: [integer]`, that war transformed as ` max_string` as a slot name. If we have focus on final user, we can change the names

Okay, then let's change them, but maybe then characters_max and characters_min, because string is rather a programming thing. If I just think about words, I would say character I guess. In german Zeichen, right?

> 
> > data_allow_paste: is this a qti feature or opal?
> 
> yes, it is only for opal. I should have created inherited class EssayOpal, but only due to one slot - i didnt) should i do it?

I see. How about displaying a message if somebody uses it? Maybe even a warning.

**Comment by shevandrin** [comment]
> As an instructor I write the perfect answer in the feedback section. From this answer we can calculate expected length, maybe feedback-length * 1.5, and max as feedback-length*2. This is useful so that students know how much they need to write and to prevent too long answers.

the idea is cool. i really like it.
But the 'perfect answer' is too subjective. All people have different style of writing (some do it too wordy), you can explain some concept using five words, other need more))  it is too harsh, unless your ask a very specific question.

**Comment by johannes-titz** [comment]
Of course you are right. But I think this still makes for better defaults than just putting fixed numbers. Sometimes in previous exams we explicitly asked students to be concise or to not write more than x words. And if the instructor is not happy, he can always adjust the parameters.

**Comment by shevandrin** [comment]
> Okay, then let's change them, but maybe then characters_max and characters_min, because string is rather a programming thing. If I just think about words, I would say character I guess. In german Zeichen, right?

it counts not characters, rather words, so can it be like `words_max`?

**Comment by johannes-titz** [comment]
Oh really? Ok, then words_max is much  much better than strings_max :).

**Comment by shevandrin** [comment]
> I see. How about displaying a message if somebody uses it? Maybe even a warning.

so our compromize here is not to crate a new class, but warning that allow_data_paste works only on Opal? (i will check the Olat)

**Comment by shevandrin** [comment]
lets finalized the list of changes:

- [ ] feedback+ or feedback- for Essay give an Error
- [ ] max_strings -> words_max  and strings_min ->words_min
- [ ] data_allow_paste  with warning that it works only on Opal
- [ ] default value for words_max is number of words in model answer from feedback * 2

**Comment by bondarenko-petr** [comment]
The average word length **in the German language** is **11 characters**, while **in English**, it's **6 characters**. On the other hand, I conducted research to determine the optimal field length for inputting an answer, which is 150 characters.

Consequently, I can propose the following algorithm for calculating expected_length and expected_lines:

In the initial step, we need to determine the number of words in the feedback. This will be represented by the variable 'n_words.' 

To establish the maximum number of words, we can use the formula: 
**word_max = 2 * n_words.**

Next, we calculate the character count for 'word_max':
   - For German words: **n_characters = 11 * words_max**
   - For English words: **n_characters = 6 * words_max**

Now, we can calculate the variables 'expected_length' and 'expected_lines' as follows:

If **n_characters is less than 150**, the variables will have the following values:
     - **expected_length = n_characters**
     - **expected_lines = 1**

If **n_characters exceed 150**, the variables will take on the following values:
     - **expected_length = 150**
     - **expected_lines = round(n_characters / 150) + 1**

Example: nwords = 45, word_max = 90, n_character = 90*6=540, expected_lenght = 150, expected_line = 5
![image](https://github.com/shevandrin/qti/assets/117097478/d9d3ecd3-558a-4a4a-b6d1-ecbe8cdf9ef1)

Example: nwords = 5, word_max = 10, n_character = 10*6 =60, expected_lenght = 60, expected_line = 1
![image](https://github.com/shevandrin/qti/assets/117097478/cb965a22-aa4f-46de-bd10-1f4941393134)
